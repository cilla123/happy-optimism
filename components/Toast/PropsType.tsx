export default interface PropsType {
  visible?: boolean;
  stayTime?: number;
  onClose?: () => void;
  mask?: boolean;
  onMaskClick?: () => void;
}

declare global {
  interface Window {
    zarmToast?: any;
  }
}
