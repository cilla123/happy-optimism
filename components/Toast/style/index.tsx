import '../../style/';
import '../../Mask/style/';
import './index.scss';
