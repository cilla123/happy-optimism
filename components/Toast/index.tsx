import React, { PureComponent } from 'react';
import ReactDOM from 'react-dom';
import classnames from 'classnames';
import PropsType from './PropsType';
import Mask from '../Mask';

export interface ToastProps extends PropsType {
  prefixCls?: string;
  className?: string;
}

export default class Toast extends PureComponent<ToastProps, any> {

  static defaultProps = {
    prefixCls: 'za-toast',
    visible: false,
    stayTime: 3000,
    mask: false,
  };

  static show = (children: any, stayTime?: number, onClose?: () => void) => {
    ReactDOM.render(
      <Toast visible stayTime={stayTime} onClose={onClose}>{children}</Toast>
    , window.zarmToast);
  }

  static hide = () => {
    ReactDOM.render(<Toast visible={false} />, window.zarmToast);
  }

  private timer: number;

  constructor(props) {
    super(props);
    this.state = {
      visible: props.visible || false,
    };
  }

  componentDidMount() {
    if (this.props.visible) {
      this.enter(this.props);
    }
  }

  componentWillReceiveProps(nextProps) {
    clearTimeout(this.timer);

    if (nextProps.visible) {
      this.enter(nextProps);
    } else {
      this.leave();
    }
  }

  componentWillUnmount() {
    clearTimeout(this.timer);
  }

  enter = (props) => {
    const { stayTime, onMaskClick } = props;

    this.setState({
      visible: true,
    });

    if (stayTime === 0) {
      return;
    }

    this.timer = setTimeout(() => {
      if (typeof onMaskClick === 'function') {
        onMaskClick();
      }
      this.leave();
      clearTimeout(this.timer);
    }, stayTime);
  }

  leave = () => {
    this.setState({
      visible: false,
    });

    const { onClose } = this.props;
    if (typeof onClose === 'function') {
      onClose();
    }
  }

  render() {
    const { prefixCls, className, mask, onMaskClick, children } = this.props;
    const { visible } = this.state;

    const cls = classnames(`${prefixCls}`, className, {
      [`${prefixCls}-open`]: visible,
    });

    return (
      <div className={cls}>
        <div className={`${prefixCls}-container`}>
          {children}
        </div>
        {mask && <Mask type="transparent" onClose={onMaskClick} />}
      </div>
    );
  }
}

if (!window.zarmToast) {
  window.zarmToast = document.createElement('div');
  document.body.appendChild(window.zarmToast);
}

ReactDOM.render(<Toast visible={false} />, window.zarmToast);
