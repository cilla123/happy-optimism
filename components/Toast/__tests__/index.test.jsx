import React from 'react';
import { render, shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Toast from '../index';

describe('Toast', () => {
  it('renders correctly', () => {
    const wrapper = render(
      <Toast>foo</Toast>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('visible change false', () => {
    const wrapper = mount(
      <Toast visible>foo</Toast>
    );
    wrapper.setProps({ visible: false });
  });

  it('stayTime', () => {
    jest.useFakeTimers();
    const wrapper = mount(
      <Toast stayTime={5000}>foo</Toast>
    );
    wrapper.setProps({ visible: false });
    jest.runAllTimers();
    wrapper.unmount();
  });

  it('stayTime is 0', () => {
    const wrapper = shallow(
      <Toast stayTime={0}>foo</Toast>
    );
    wrapper.setProps({ visible: true });
  });
});
