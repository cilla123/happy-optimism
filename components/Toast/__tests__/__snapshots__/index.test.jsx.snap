// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toast renders correctly 1`] = `
<div
  class="za-toast"
>
  <div
    class="za-toast-container"
  >
    foo
  </div>
</div>
`;
